File: domain.py
import requests
from bs4 import BeautifulSoup
import json
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time

headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
    "Accept-Language": "en-US,en;q=0.5",
    "Referer": "https://www.google.com/"
}

def get_final_url_selenium(url, retries=3, wait_time=10):
    """Use Selenium to handle JavaScript-based redirects and return the final URL."""
    for attempt in range(retries):
        try:
            options = Options()
            options.add_argument("--headless")  # Run in headless mode
            options.add_argument(f"user-agent={headers['User-Agent']}")
            options.add_argument("--disable-gpu")
            options.add_argument("--no-sandbox")
            options.add_argument("--disable-dev-shm-usage")
            options.add_argument("--ignore-certificate-errors")
            options.add_argument("--disable-popup-blocking")  # Allow new tabs
            options.add_argument("--disable-blink-features=AutomationControlled")  # Avoid bot detection
            
            driver = webdriver.Chrome(options=options)
            driver.get(url)
            time.sleep(3)  # Wait for initial redirects
            
            final_url = driver.current_url
            print(f"📍 Initial URL after loading {url}: {final_url}")
            
            # Handle HDHub4u-specific redirects
            if "hdhub4u.mn" in final_url:
                try:
                    # Wait for the "View Full Site" button to be clickable
                    button = WebDriverWait(driver, wait_time).until(
                        EC.element_to_be_clickable((By.CSS_SELECTOR, "a.new-tab.btn.btn-lg.btn-radius.btn-primary"))
                    )
                    # Store the original window handle
                    original_window = driver.current_window_handle
                    button.click()
                    time.sleep(7)  # Wait for new tab and redirects
                    
                    # Switch to the new tab (if opened)
                    new_tab_opened = False
                    for window_handle in driver.window_handles:
                        if window_handle != original_window:
                            driver.switch_to.window(window_handle)
                            new_tab_opened = True
                            time.sleep(5)  # Wait for redirects in new tab
                            final_url = driver.current_url
                            print(f"✅ New tab URL after clicking 'View Full Site' for {url}: {final_url}")
                            break
                    
                    # If no new tab or still on hdhub4u.mn, check "Click Here" link
                    if not new_tab_opened or "hdhub4u.mn" in final_url:
                        try:
                            stx_link = WebDriverWait(driver, wait_time).until(
                                EC.presence_of_element_located((By.CSS_SELECTOR, "#stx a"))
                            )
                            href = stx_link.get_attribute("href")
                            if href:
                                print(f"📌 Found 'Click Here' link in #stx: {href}")
                                driver.get(href)
                                time.sleep(7)  # Wait for redirects
                                final_url = driver.current_url
                                print(f"✅ URL after following 'Click Here' for {url}: {final_url}")
                        except Exception as e:
                            print(f"❌ No 'Click Here' link found in #stx or error: {e}")
                except Exception as e:
                    print(f"❌ Failed to click 'View Full Site' for {url}: {e}")
                    # Fallback to "Click Here" link
                    try:
                        stx_link = WebDriverWait(driver, wait_time).until(
                            EC.presence_of_element_located((By.CSS_SELECTOR, "#stx a"))
                        )
                        href = stx_link.get_attribute("href")
                        if href:
                            print(f"📌 Found 'Click Here' link in #stx: {href}")
                            driver.get(href)
                            time.sleep(7)  # Wait for redirects
                            final_url = driver.current_url
                            print(f"✅ URL after following 'Click Here' for {url}: {final_url}")
                    except Exception as e:
                        print(f"❌ No 'Click Here' link found in #stx or error: {e}")
            else:
                # For KatWorld or other URLs, wait for JavaScript redirects
                time.sleep(wait_time)
                final_url = driver.current_url
                print(f"✅ Selenium final URL for {url}: {final_url}")
            
            driver.quit()
            
            # Verify final URL with requests to follow redirects
            try:
                response = requests.get(final_url, headers=headers, allow_redirects=True, timeout=10)
                final_url = response.url
                print(f"✅ Verified final URL with requests: {final_url}")
            except Exception as e:
                print(f"⚠️ Failed to verify final URL with requests: {e}")
            
            return final_url
        except Exception as e:
            print(f"❌ Selenium error for {url} (attempt {attempt + 1}/{retries}): {e}")
            if attempt == retries - 1:
                driver.quit()
                return url
            time.sleep(2)  # Wait before retrying

def extract_katworld_links():
    url = "https://katworld.net/"
    try:
        r = requests.get(url, headers=headers, timeout=10)
        soup = BeautifulSoup(r.text, "html.parser")

        links = {
            "hollywood": "",
            "anime": "",
            "4k": "",
            "adult": "",
            "drama": ""
        }

        for a in soup.find_all("a", href=True):
            href = a["href"]
            text = a.get_text(strip=True).lower()

            if "katmoviehd" in text:
                print(f"📌 Found KatWorld hollywood link: {href}")
                links["hollywood"] = get_final_url_selenium(href)
            elif "pikahd" in text or "anime" in text:
                print(f"📌 Found KatWorld anime link: {href}")
                links["anime"] = get_final_url_selenium(href)
            elif "katmovie4k" in text or "4k" in text:
                print(f"📌 Found KatWorld 4k link: {href}")
                links["4k"] = get_final_url_selenium(href)
            elif "katmovie18" in text or "adult" in text:
                print(f"📌 Found KatWorld adult link: {href}")
                links["adult"] = get_final_url_selenium(href)
            elif "katdrama" in text or "drama" in text:
                print(f"📌 Found KatWorld drama link: {href}")
                links["drama"] = get_final_url_selenium(href)

        return links
    except Exception as e:
        print("❌ KatWorld error:", e)
        return {}

def extract_hdhub4u_main_link():
    url = "https://hdhublist.com/"
    try:
        r = requests.get(url, headers=headers, timeout=10)
        soup = BeautifulSoup(r.text, "html.parser")

        for a in soup.find_all("a", href=True):
            text = a.get_text(strip=True).lower()
            if "hdhub4u" in text and "main site" in text:
                href = a["href"]
                if href.startswith("/"):
                    href = f"https://hdhublist.com{href}"
                print(f"📌 Found HDHub4u link: {href}")
                return get_final_url_selenium(href)
        
        print("⚠️ No 'main site' link found, using fallback URL")
        return get_final_url_selenium("https://hdhub4u.tv/")
    except Exception as e:
        print("❌ HDHub4u error:", e)
        return get_final_url_selenium("https://hdhub4u.tv/")

def update_domains_json():
    katworld_links = extract_katworld_links()
    hdhub4u_main = extract_hdhub4u_main_link()

    domains = {
        "katworld": katworld_links,
        "hdhub4u": {
            "main": hdhub4u_main
        }
    }

    with open("domains.json", "w") as f:
        json.dump(domains, f, indent=4)

    print("✅ domains.json updated:\n")
    print(json.dumps(domains, indent=4))

if __name__ == "__main__":
    update_domains_json()

File: domain.js
const axios = require('axios');
const cheerio = require('cheerio');
const fs = require('fs').promises;
const puppeteer = require('puppeteer');

// User agent for requests
const headers = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36',
  'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
  'Accept-Language': 'en-US,en;q=0.5',
  'Referer': 'https://www.google.com/'
};

/**
 * Use Puppeteer to handle JavaScript-based redirects and return the final URL
 * @param {string} url - The initial URL to navigate to
 * @param {number} retries - Number of retry attempts
 * @param {number} waitTime - Time to wait for redirects in seconds
 * @returns {Promise<string>} The final URL after all redirects
 */
async function getFinalUrlSelenium(url, retries = 3, waitTime = 10) {
  let browser = null;
  
  for (let attempt = 0; attempt < retries; attempt++) {
    try {
      // Launch browser
      browser = await puppeteer.launch({
        headless: true,
        args: [
          '--no-sandbox',
          '--disable-gpu',
          '--disable-dev-shm-usage',
          '--ignore-certificate-errors',
          '--disable-popup-blocking',
          '--disable-blink-features=AutomationControlled'
        ]
      });
      
      const page = await browser.newPage();
      
      // Set user agent
      await page.setUserAgent(headers['User-Agent']);
      
      // Navigate to URL
      await page.goto(url, { waitUntil: 'networkidle2', timeout: 30000 });
      // Use setTimeout instead of waitForTimeout
      await new Promise(resolve => setTimeout(resolve, 3000)); // Wait for initial redirects
      
      let finalUrl = page.url();
      console.log(`📍 Initial URL after loading ${url}: ${finalUrl}`);
      
      // Handle HDHub4u-specific redirects
      if (finalUrl.includes('hdhub4u.do')) {
        try {
          // Wait for the "View Full Site" button to be clickable
          const button = await page.waitForSelector('a.new-tab.btn.btn-lg.btn-radius.btn-primary', { timeout: waitTime * 1000 });
          
          // Store the original page
          const originalPage = page;
          
          // Create a promise to wait for a new page to open
          const newPagePromise = new Promise(resolve => {
            browser.once('targetcreated', async target => {
              const newPage = await target.page();
              resolve(newPage);
            });
          });
          
          // Click the button
          await button.click();
          
          // Wait for the new page to open
          const newPage = await newPagePromise;
          // Use setTimeout instead of waitForTimeout
          await new Promise(resolve => setTimeout(resolve, 7000)); // Wait for redirects in new tab
          
          finalUrl = newPage.url();
          console.log(`✅ New tab URL after clicking 'View Full Site' for ${url}: ${finalUrl}`);
          
          // If still on hdhub4u.mn, check "Click Here" link
          if (finalUrl.includes('hdhub4u.mn')) {
            try {
              const stxLink = await newPage.waitForSelector('#stx a', { timeout: waitTime * 1000 });
              const href = await stxLink.evaluate(el => el.getAttribute('href'));
              
              if (href) {
                console.log(`📌 Found 'Click Here' link in #stx: ${href}`);
                await newPage.goto(href, { waitUntil: 'networkidle2', timeout: 30000 });
                // Use setTimeout instead of waitForTimeout
                await new Promise(resolve => setTimeout(resolve, 7000)); // Wait for redirects
                finalUrl = newPage.url();
                console.log(`✅ URL after following 'Click Here' for ${url}: ${finalUrl}`);
              }
            } catch (e) {
              console.log(`❌ No 'Click Here' link found in #stx or error: ${e.message}`);
            }
          }
        } catch (e) {
          console.log(`❌ Failed to click 'View Full Site' for ${url}: ${e.message}`);
          
          // Fallback to "Click Here" link
          try {
            const stxLink = await page.waitForSelector('#stx a', { timeout: waitTime * 1000 });
            const href = await stxLink.evaluate(el => el.getAttribute('href'));
            
            if (href) {
              console.log(`📌 Found 'Click Here' link in #stx: ${href}`);
              await page.goto(href, { waitUntil: 'networkidle2', timeout: 30000 });
              // Use setTimeout instead of waitForTimeout
              await new Promise(resolve => setTimeout(resolve, 7000)); // Wait for redirects
              finalUrl = page.url();
              console.log(`✅ URL after following 'Click Here' for ${url}: ${finalUrl}`);
            }
          } catch (e) {
            console.log(`❌ No 'Click Here' link found in #stx or error: ${e.message}`);
          }
        }
      } else {
        // For KatWorld or other URLs, wait for JavaScript redirects
        // Use setTimeout instead of waitForTimeout
        await new Promise(resolve => setTimeout(resolve, waitTime * 1000));
        finalUrl = page.url();
        console.log(`✅ Puppeteer final URL for ${url}: ${finalUrl}`);
      }
      
      await browser.close();
      browser = null;
      
      // Verify final URL with axios to follow redirects
      try {
        const response = await axios.get(finalUrl, { 
          headers, 
          maxRedirects: 5,
          timeout: 10000
        });
        finalUrl = response.request.res.responseUrl || finalUrl;
        console.log(`✅ Verified final URL with axios: ${finalUrl}`);
      } catch (e) {
        console.log(`⚠️ Failed to verify final URL with axios: ${e.message}`);
      }
      
      return finalUrl;
    } catch (e) {
      console.log(`❌ Puppeteer error for ${url} (attempt ${attempt + 1}/${retries}): ${e.message}`);
      
      if (browser) {
        await browser.close();
        browser = null;
      }
      
      if (attempt === retries - 1) {
        return url;
      }
      
      // Wait before retrying
      await new Promise(resolve => setTimeout(resolve, 2000));
    }
  }
  
  return url;
}

/**
 * Extract KatWorld links from their main page
 * @returns {Promise<Object>} Object containing KatWorld links
 */
async function extractKatworldLinks() {
  const url = 'https://katworld.net/';
  
  try {
    const response = await axios.get(url, { headers, timeout: 10000 });
    const $ = cheerio.load(response.data);
    
    const links = {
      'hollywood': '',
      'anime': '',
      '4k': '',
      'adult': '',
      'drama': ''
    };
    
    // Find all links and collect them first
    const foundLinks = [];
    
    $('a').each((_, element) => {
      const href = $(element).attr('href');
      const text = $(element).text().toLowerCase();
      
      if (href) {
        if (text.includes('katmoviehd')) {
          console.log(`📌 Found KatWorld hollywood link: ${href}`);
          foundLinks.push({ type: 'hollywood', href });
        } else if (text.includes('pikahd') || text.includes('anime')) {
          console.log(`📌 Found KatWorld anime link: ${href}`);
          foundLinks.push({ type: 'anime', href });
        } else if (text.includes('katmovie4k') || text.includes('4k')) {
          console.log(`📌 Found KatWorld 4k link: ${href}`);
          foundLinks.push({ type: '4k', href });
        } else if (text.includes('katmovie18') || text.includes('adult')) {
          console.log(`📌 Found KatWorld adult link: ${href}`);
          foundLinks.push({ type: 'adult', href });
        } else if (text.includes('katdrama') || text.includes('drama')) {
          console.log(`📌 Found KatWorld drama link: ${href}`);
          foundLinks.push({ type: 'drama', href });
        }
      }
    });
    
    // Process links sequentially
    for (const link of foundLinks) {
      try {
        links[link.type] = await getFinalUrlSelenium(link.href);
      } catch (error) {
        console.log(`❌ Error processing ${link.type} link: ${error.message}`);
      }
    }
    
    return links;
  } catch (error) {
    console.log(`❌ KatWorld error: ${error.message}`);
    return {};
  }
}

/**
 * Extract HDHub4u main link
 * @returns {Promise<string>} The HDHub4u main link
 */
async function extractHdhub4uMainLink() {
  const url = 'https://hdhublist.com/';
  
  try {
    const response = await axios.get(url, { headers, timeout: 10000 });
    const $ = cheerio.load(response.data);
    
    let mainLink = null;
    
    // Find the HDHub4u link
    $('a').each((_, element) => {
      const text = $(element).text().toLowerCase();
      if (text.includes('hdhub4u') && text.includes('main site')) {
        let href = $(element).attr('href');
        if (href && href.startsWith('/')) {
          href = `https://hdhublist.com${href}`;
        }
        console.log(`📌 Found HDHub4u link: ${href}`);
        mainLink = href;
      }
    });
    
    // Process the link
    try {
      if (mainLink) {
        return await getFinalUrlSelenium(mainLink);
      } else {
        console.log('⚠️ No "main site" link found, using fallback URL');
        return await getFinalUrlSelenium('https://hdhub4u.tv/');
      }
    } catch (error) {
      console.log(`❌ Error processing HDHub4u link: ${error.message}`);
      return 'https://hdhub4u.tv/';
    }
  } catch (error) {
    console.log(`❌ HDHub4u error: ${error.message}`);
    return await getFinalUrlSelenium('https://hdhub4u.tv/');
  }
}

/**
 * Update domains.json with the latest working domains
 * @returns {Promise<void>}
 */
async function updateDomainsJson() {
  try {
    // Read existing domains.json first as a fallback
    let existingDomains = {};
    try {
      const existingData = await fs.readFile('domains.json', 'utf8');
      existingDomains = JSON.parse(existingData);
    } catch (err) {
      console.log('No existing domains.json found or error reading it');
    }
    
    // Try to get new links
    let katworldLinks = {};
    let hdhub4uMain = '';
    
    try {
      katworldLinks = await extractKatworldLinks();
    } catch (err) {
      console.error(`Error extracting KatWorld links: ${err.message}`);
      // Use existing values as fallback
      if (existingDomains.katworld) {
        katworldLinks = existingDomains.katworld;
      }
    }
    
    try {
      hdhub4uMain = await extractHdhub4uMainLink();
    } catch (err) {
      console.error(`Error extracting HDHub4u link: ${err.message}`);
      // Use existing value as fallback
      if (existingDomains.hdhub4u && existingDomains.hdhub4u.main) {
        hdhub4uMain = existingDomains.hdhub4u.main;
      }
    }
    
    // Create new domains object
    const domains = {
      'katworld': katworldLinks,
      'hdhub4u': {
        'main': hdhub4uMain
      }
    };
    
    // Write to file
    await fs.writeFile('domains.json', JSON.stringify(domains, null, 4));
    
    console.log('✅ domains.json updated:\n');
    console.log(JSON.stringify(domains, null, 4));
    
    return domains;
  } catch (error) {
    console.error(`Error updating domains.json: ${error.message}`);
    throw error;
  }
}

// If this file is run directly, update domains.json
if (require.main === module) {
  updateDomainsJson();
}

module.exports = {
  getFinalUrlSelenium,
  extractKatworldLinks,
  extractHdhub4uMainLink,
  updateDomainsJson
};